### PR TITLE
Implement scheduled scans

### PR DESCRIPTION
**Pull request by ro-tex**
# PULL REQUEST

## Overview

**TL;DR:**
This is a poor man's balancing of pinned files. All servers are scheduled to scan for underpinned files at the same time, so the load is spread.

**Full story**
After a server is [marked as dead](https://linear.app/skynetlabs/issue/SKY-297/implement-an-endpoint-which-marks-a-given-server-as-dead), all of the skylinks it used to pin would be underpinned. These should be repinned as soon as possible by other servers. We want those files spread among all servers in the cluster and not just pinned on the first server which happens to perform a scan but we still don't have the [load balancing of pinned files](https://linear.app/skynetlabs/issue/SKY-1078/load-balancing-pinned-files) implemented. That's why we'll do a poor man's version of it and we'll schedule all servers to perform their scans at the same time, giving each of them a chance to grab some files.

This PR depends on #38 and #40. Once they are merged, it will be significantly easier to review. If you want to review before that, you can check commits 0fef2a0e1391f26e551255777ac7b584a28596e0 and the one after.

## Checklist

- [ ] All git commits are signed. (REQUIRED)
- [ ] All new methods or updated methods have clear docstrings.
- [ ] Testing added or updated for new methods.
- [ ] Verified if any changes impact the WebPortal Health Checks.
- [ ] Appropriate documentation updated.
- [ ] Changelog file created.

## Issues Closed

Closes SKY-1250

### CODE REVIEW

**Comment by peterjan** [comment]
pretty close to done I'd say, nice 👍

File: conf/configuration.go [hunk 1]
```diff
@@ -203,9 +229,48 @@ func MinPinners(ctx context.Context, db *database.DB) (int, error) {
...
+func NextScan(ctx context.Context, db *database.DB, logger logger.Logger) (time.Time, error) {
+	val, err := db.ConfigValue(ctx, ConfNextScan)
+	if errors.Contains(err, mongo.ErrNoDocuments) {
+		logger.Infof("Missing database value for '%s', setting a new one.", ConfNextScan)
+		// No scan has been scheduled. Schedule one in an hour.
+		err = SetNextScan(ctx, db, time.Now().Add(DefaultNextScanOffset).UTC())
+		if err != nil {
+			return time.Time{}, err
+		}
+		// Read the value from the DB in order to avoid race conditions.
```
[the remark this answers is not in the document]
Comment (ro-tex) [reply conf/configuration.go:249]: The idea was that another server could be doing exactly the same thing at the same time and it's also trying to set the time of the next scan, potentially overwriting the value we set here. I wanted to have the exact time of the run, which would require a DB read.
After I read it again, the two values set by the two servers will be close enough that it doesn't really matter, so we don't need to read from the DB after all. I changed the code.

File: conf/configuration.go [hunk 2]
```diff
@@ -203,9 +229,48 @@ func MinPinners(ctx context.Context, db *database.DB) (int, error) {
...
+		// While this does not eliminate data races, it does help.
+		return NextScan(ctx, db, logger)
```
[the remark this answers is not in the document]
Comment (ro-tex) [reply conf/configuration.go:251]: The only way this is an endless loop is if MongoDB fails to insert a record in the DB *and* erroneously returns "no error". Similarly, the next `NextScan` recursive call can only be endless if MongoDB fails to write a valid RFC3339 time value somehow. Both situations are unlikely and depend of severe bugs in Mongo, which is not something we should reasonably code around.

In any case, this is moot because the recursive `NextScan` calls are removed.

File: conf/configuration.go [hunk 3]
```diff
@@ -203,9 +229,48 @@ func MinPinners(ctx context.Context, db *database.DB) (int, error) {
...
+	}
+	if err != nil {
+		return time.Time{}, err
+	}
+	t, err := time.Parse(time.RFC3339, val)
+	if err != nil {
+		logger.Warnf("Invalid database value for '%s': '%s', setting a new one.", ConfNextScan, val)
```
[the remark this answers is not in the document]
Comment (ro-tex) [reply conf/configuration.go:258]: Hmm... I will actually elevate it to `build.Critical` because this should never happen. I was just wrapping edge cases here but if this ever happens, it's likely a programmer error somewhere.

File: conf/configuration.go [hunk 4]
```diff
@@ -203,9 +229,48 @@ func MinPinners(ctx context.Context, db *database.DB) (int, error) {
...
+		err = SetNextScan(ctx, db, time.Now().Add(DefaultNextScanOffset).UTC())
+		if err != nil {
+			return time.Time{}, err
+		}
+		// Read the value from the DB in order to avoid race conditions.
```
[the remark this answers is not in the document]
Comment (ro-tex) [reply conf/configuration.go:263]: As I said in https://github.com/SkynetLabs/pinner/pull/41#discussion_r934246666, I don't think this data race is important in any way. It can only happen if:
1. we are just deploying this functionality for the first time on that cluster
2. we have two servers which have different builds
3. we have changed the value of `DefaultNextScanOffset` between those builds

Even if all those are somehow met, the scan also needs to be within the `SleepBetweenChecksForScan` period for both servers in order to cause an actual discrepancy. And even if all those thing happen, the end result will still be that one server start scanning for underpinned skylinks a bit before the other. That will have zero effect on a brand new cluster and a very small effect on an existing cluster because we intentionally pause between skylink repins.

All in all, I don't think it's worth it to add the extra complexity and DB load of using locks. I changed this code to 
```go
		scanTime := time.Now().Add(DefaultNextScanOffset).UTC()
		err = SetNextScan(ctx, db, scanTime)
		if err != nil {
			return time.Time{}, err
		}
		return scanTime, nil
```
based on your previous comment and I think that's the right call.

File: conf/configuration.go [hunk 5]
```diff
@@ -203,9 +229,48 @@ func MinPinners(ctx context.Context, db *database.DB) (int, error) {
...
+	}
+	return t, nil
+}
+
+// SetNextScan sets the time of the next cluster-wide scan for underpinned files.
+func SetNextScan(ctx context.Context, db *database.DB, t time.Time) error {
+	if t.Before(time.Now().UTC().Add(SleepBetweenChecksForScan)) {
+		return ErrTimeTooSoon
+	}
+	return db.SetConfigValue(ctx, ConfNextScan, t.UTC().Format(time.RFC3339))
```
[the remark this answers is not in the document]
Comment (ro-tex) [reply conf/configuration.go:275]: That's a smart thing to do, indeed. Thanks!

File: test/conf/configuration_test.go [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+// function.
+func curryDB(fn func(t *testing.T, db *database.DB), db *database.DB) func(t *testing.T) {
+	return func(t *testing.T) {
+		fn(t, db)
+	}
+}
+
+// testDryRun ensures DryRun functions correctly.
+func testDryRun(t *testing.T, db *database.DB) {
+	ctx := context.Background()
```
[the remark this answers is not in the document]
Comment (ro-tex) [reply test/conf/configuration_test.go:51]: Added.

File: test/conf/configuration_test.go [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+	if !errors.Contains(err, mongo.ErrNoDocuments) {
+		t.Fatalf("Expected '%s', got '%s'", mongo.ErrNoDocuments, err)
+	}
+	// Get the NextScan value, expect this to set the default offset.
+	t0, err := conf.NextScan(ctx, db, logger)
+	if err != nil {
+		t.Fatal(err)
+	}
+	now := time.Now().UTC()
+	if t0.After(now.Add(time.Second)) || t0.Before(now.Add(-1*time.Second)) {
```
[the remark this answers is not in the document]
Comment (ro-tex) [reply test/conf/configuration_test.go:170]: Essentially, you're suggesting to change the tolerance value from 1s to 5s here, so slow DB won't cause an NDF?
Sure thing, done.

File: test/conf/configuration_test.go [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+		t.Fatalf("Expected '%s', got '%s'", now.Add(conf.DefaultNextScanOffset), t0)
+	}
+	// Try to set an invalid value.
+	err = conf.SetNextScan(ctx, db, time.Now().UTC().Add(-1*time.Hour))
```
[the remark this answers is not in the document]
Comment (ro-tex) [reply test/conf/configuration_test.go:174]: Yes, it will. But this one is really obvious when reading the test. Do you think `time.Now().UTC()` gives us anything extra? If we want to have the test value closer to the limit maybe `time.Now().UTC().Add(SleepBetweenChecksForScan - time.Millisecond)` will be better, as it cuts it right to the limit?

**Review by ChrisSchinnerl — APPROVED** [review]

**Review by ChrisSchinnerl — DISMISSED** [review]

File: conf/configuration.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (peterjan) [reply conf/configuration.go]: works for me

File: conf/configuration.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (peterjan) [reply conf/configuration.go]: oh yeah as long as we get rid of the comment that mentions the race condition as well as the redundant fetch I'm totally fine, thanks!

File: test/conf/configuration_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (peterjan) [reply test/conf/configuration_test.go]: just any test that asserts a time between x and y is always going to NDF at some point, thanks for updating

File: test/conf/configuration_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (peterjan) [reply test/conf/configuration_test.go:178]: nah fine to resolve

**Review by peterjan — APPROVED** [review]

**Review by ChrisSchinnerl — APPROVED** [review]
